src/ILCompiler.Compiler/src/Compiler/MultiFileCompilationModuleGroup.cs

```diff
@@ -34,53 +39,55 @@ public override bool ContainsType(TypeDesc type)
             return true;
         }
 
-        public override bool ContainsMethod(MethodDesc method)
+        public sealed override bool ContainsMethod(MethodDesc method)
         {
             if (method.HasInstantiation)
                 return true;
 
             return ContainsType(method.OwningType);
         }
 
-        private bool BuildingLibrary
-        {
-            get
-            {
-                foreach (var module in _compilationModuleSet)
-                {
-                    if (module.PEReader.PEHeaders.IsExe)
-                    {
-                        return false;
-                    }
-                }
-                return true;
-            }
-        }
-
         private bool IsModuleInCompilationGroup(EcmaModule module)
         {
             return _compilationModuleSet.Contains(module);
         }
 
-        public override bool IsSingleFileCompilation
+        public sealed override bool IsSingleFileCompilation
         {
             get
             {
                 return false;
             }
         }
 
+        public sealed override bool ShouldReferenceThroughImportTable(TypeDesc type)
+        {
+            return false;
+        }
+    }
+
+    public class MultiFileLibraryCompilationModuleGroup : MultiFileCompilationModuleGroup
+    {
+        public MultiFileLibraryCompilationModuleGroup(TypeSystemContext context, IEnumerable<ModuleDesc> compilationModuleSet)
+            : base(context, compilationModuleSet)
+        {
+        }
+
         public override bool ShouldProduceFullType(TypeDesc type)
         {
-            // TODO: Remove this once we have delgate constructor transform added and GetMethods() tells us about
-            //       the virtuals we add on to delegate types.
-            if (type.IsDelegate)
-                return false;
+            return true;
+        }
+    }
 
-            // Fully build all types when building a library
-            if (BuildingLibrary)
-                return true;
+    public class MultiFileLeafCompilationModuleGroup : MultiFileCompilationModuleGroup
```

Comment:
Would you mind adding comments for the various multifile groups so we remember. The naming doesn't make it super obvious and I can't think of anything amazingly better than Leaf. (Application? Entrypoint?)